cmd/immuadmin/command/hot_backup.go

```diff
@@ -0,0 +1,661 @@
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "Error switching off replica mode for db: %v", err)
+				}
+			}()
+
+			return cl.runHotRestore(file, params.progress, firstTx, signature, payload)
+		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			verify, _ := cmd.Flags().GetBool("verify")
+			if verify {
+				return cobra.ExactArgs(0)(cmd, args)
+			}
+			return cobra.ExactArgs(1)(cmd, args)
+		},
+	}
+	ccmd.Flags().StringP("input", "i", "-", "input file file, \"-\" for stdin")
+	ccmd.Flags().Bool("verify", false, "do not restore data, only verify backup")
+	ccmd.Flags().Bool("append", false, "appending to DB, if it already exists")
+	ccmd.Flags().Bool("progress", false, "show progress indicator")
+	ccmd.Flags().Bool("force", false, "don't check transaction sequence")
+	cmd.AddCommand(ccmd)
+}
+
+func prepareRestoreParams(flags *pflag.FlagSet) (*restoreParams, error) {
+	var params restoreParams
+	var err error
+
+	params.input, err = flags.GetString("input")
+	if err != nil {
+		return nil, err
+	}
+	params.append, err = flags.GetBool("append")
+	if err != nil {
+		return nil, err
+	}
+	params.force, err = flags.GetBool("force")
+	if err != nil {
+		return nil, err
+	}
+	params.verify, err = flags.GetBool("verify")
+	if err != nil {
+		return nil, err
+	}
+	params.progress, err = flags.GetBool("progress")
+	if err != nil {
+		return nil, err
+	}
+
+	return &params, nil
+}
+
+func verifyFile(file io.Reader) error {
+	fileStart, _, _, err := getTx(file)
+	if err != nil {
+		return err
+	}
+
+	last := fileStart
+	last, _, err = lastTxInFile(file)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Backup file contains transactions %d - %d\n", fileStart, last)
+	return nil
+}
+
+func (cl *commandlineHotBck) isDbExists(name string) (bool, error) {
+	dbList, err := cl.immuClient.DatabaseList(cl.context)
+	if err != nil {
+		return false, err
+	}
+
+	dbExist := false
+	for _, db := range dbList.Databases {
+		if db.DatabaseName == name {
+			dbExist = true
+			break
+		}
+	}
+
+	return dbExist, nil
+}
+
+func (cl *commandlineHotBck) verifyDb(params *restoreParams, name string, file io.Reader) (uint64, []byte, []byte, error) {
+	udr, err := cl.immuClient.UseDatabase(cl.context, &schema.Database{DatabaseName: name})
+	if err != nil {
+		return 0, nil, nil, err
+	}
+	cl.context = metadata.NewOutgoingContext(cl.context, metadata.Pairs("authorization", udr.GetToken()))
+
+	state, err := cl.immuClient.CurrentState(cl.context)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	if state.TxId > 0 {
+		// database isn't empty
+		if !params.append {
+			return 0, nil, nil, errors.New("cannot restore to non-empty database without --append flag")
+		}
+		// find the nearest transaction in backup file and position the file just after the transaction
+		txId, signature, payload, err := findTx(file, state.TxId)
+		if err != nil {
+			return 0, nil, nil, err
+		}
+		gap := txId - state.TxId
```

Comment:
I'm not sure I follow the gap cases, which situations are being covered?

Reply:
There are three possible scenarios. Imagine the DB that has transactions 1-10.
1. User wants to append transactions 10-20 from file, there is an overlap (transaction 10 is in both DB and restore file), so restore process can validate that checksums for txn 10 in DB and file do match, to make sure user didn't mess up.
2. User wants to append transactions 11-20, in theory it is possible as there is no gap between transactions, but because there is no overlap, checksum cannot be verified. Therefore restore process wants `--force` param to make sure user really understands.
3. User wants to append transactions 12-20, there is a gap (transaction 11 is missing), restore not possible.

Reply:
either if there is overlap or a gap, the replication will automatically fail, validations are already in place in the store.

probably is simpler to display information about what's the latest tx in the database, and what are the available tx in the backup file?

Reply:
> either if there is overlap or a gap, the replication will automatically fail, validations are already in place in the store.

Yes, but the error message is very cryptic, user cannot understand the source of the problem. I test it in order to give better diagnostics

Reply:
understood